apps/server/src/modules/oauth/service/oauth.service.ts

```diff
@@ -29,176 +28,129 @@ export class OAuthService {
 		private readonly logger: Logger,
 		private readonly provisioningService: ProvisioningService,
 		private readonly systemService: SystemService,
-		private readonly userMigrationService: UserMigrationService
+		private readonly userMigrationService: UserMigrationService,
+		private readonly migrationCheckService: MigrationCheckService
 	) {
 		this.logger.setContext(OAuthService.name);
 	}
 
-	async authenticateUser(
-		systemId: string,
-		authCode?: string,
-		errorCode?: string
-	): Promise<{ user?: UserDO; redirect: string }> {
-		let redirect: string;
-		if (errorCode) {
-			redirect = this.createErrorRedirect(errorCode);
-			return { user: undefined, redirect };
-		}
-		if (!authCode) {
-			throw new OAuthSSOError(
+	async authenticateUser(systemId: string, authCode?: string, errorCode?: string): Promise<OAuthTokenDto> {
+		if (errorCode || !authCode) {
+			throw new UnauthorizedException(
 				'Authorization Query Object has no authorization code or error',
 				errorCode || 'sso_auth_code_step'
 			);
 		}
 
-		const system = await this.systemService.findOAuthById(systemId);
-		if (!system.id) {
-			// unreachable. System loaded from DB always has an ID
-			throw new UnauthorizedException(`System with id "${systemId}" does not exist.`);
+		const system: SystemDto = await this.systemService.findOAuthById(systemId);
+		if (!system.oauthConfig) {
+			throw new UnauthorizedException(`Requested system ${systemId} has no oauth configured`, 'sso_internal_error');
 		}
+		const { oauthConfig } = system;
 
-		const oauthConfig: OauthConfig = this.extractOauthConfigFromSystem(system);
+		const oauthTokens: OAuthTokenDto = await this.requestToken(authCode, oauthConfig, oauthConfig.redirectUri);
 
-		const queryToken: OauthTokenResponse = await this.requestToken(authCode, oauthConfig);
+		await this.validateToken(oauthTokens.idToken, oauthConfig);
 
-		await this.validateToken(queryToken.id_token, oauthConfig);
+		return oauthTokens;
+	}
 
-		const data: OauthDataDto = await this.provisioningService.getData(
-			queryToken.access_token,
-			queryToken.id_token,
-			system.id
-		);
+	async provisionUser(
+		systemId: string,
+		idToken: string,
+		accessToken: string
+	): Promise<{ user?: UserDO; redirect: string }> {
+		const data: OauthDataDto = await this.provisioningService.getData(systemId, idToken, accessToken);
 
-		// TODO Move Migration Checks to other service
+		let migrationConsentRedirect: string | undefined;
 		if (data.externalSchool?.officialSchoolNumber) {
-			const shouldMigrate: boolean = await this.shouldUserMigrate(
+			const shouldUserMigrate: boolean = await this.migrationCheckService.shouldUserMigrate(
 				data.externalUser.externalId,
-				data.externalSchool.officialSchoolNumber,
-				system.id
+				systemId,
+				data.externalSchool.officialSchoolNumber
 			);
```

Comment:
wouldn't it be better to have this call just take the whole data? OauthService shouldn't have to know what is necessary if a user is able to migrate.